pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -385,10 +387,15 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
             else:
                 checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_KEY] = dict(model.hparams)
 
-        # give the model a chance to dump a few things
+        # dump stateful datamodule
+        datamodule = self.trainer.datamodule
+        if datamodule is not None:
+            checkpoint[datamodule.__class__.__name__] = datamodule.state_dict()
```

Comment:
all `state_dict()` calls should happen before any `on_save_checkpoint` hooks are called